### PR TITLE
Add llama-cpp python support 

### PR DESCRIPTION
**Pull request by DanielUH2019**
This changes allows running some large models on CPU locally. Let me know how I can improve it

### CODE REVIEW

File: dsp/modules/cohere.py [hunk 1]
```diff
@@ -101,6 +101,9 @@ def request(self, prompt: str, **kwargs):
         """Handles retrieval of completions from Cohere whilst handling API errors"""
         return self.basic_request(prompt, **kwargs)
 
+    def get_choice_text(self, choice) -> str:
```
Comment (okhat) [inline dsp/modules/cohere.py:104]: Is this needed?

File: dsp/modules/gpt3.py [hunk 2]
```diff
@@ -117,10 +117,10 @@ def request(self, prompt: str, **kwargs) -> OpenAIObject:
         
         return self.basic_request(prompt, **kwargs)
 
-    def _get_choice_text(self, choice: dict[str, Any]) -> str:
+    def get_choice_text(self, choice: dict[str, Any]) -> str:
         if self.model_type == "chat":
-            return choice["message"]["content"]
-        return choice["text"]
+            return ' ' + choice["message"]["content"].strip()
```
Comment (okhat) [inline dsp/modules/gpt3.py:122]: Why?

File: dsp/modules/hf.py [hunk 3]
```diff
@@ -99,6 +99,9 @@ def basic_request(self, prompt, **kwargs):
         self.history.append(history)
 
         return response
+    
+    def get_choice_text(self, choice: dict[str, Any]) -> str:
+        return choice["text"]
```
Comment (okhat) [inline dsp/modules/hf.py:104]: Is this needed?

File: dsp/modules/cohere.py (hunk 1 above)
Comment (okhat) [reply dsp/modules/cohere.py:104]: oh I see, I like your refactor, this is for inspect_history

File: dsp/modules/hf.py (hunk 3 above)
Comment (okhat) [reply dsp/modules/hf.py:104]: got it

File: dsp/modules/gpt3.py [hunk 4]
```diff
@@ -157,7 +157,7 @@ def __call__(
         if only_completed and len(completed_choices):
             choices = completed_choices
 
-        completions = [self._get_choice_text(c) for c in choices]
+        completions = [self.get_choice_text(c) for c in choices]
```
Comment (okhat) [inline dsp/modules/gpt3.py:160]: This kind of changes the behavior right? It used to be "stripped", now it has space before it.

I'm not sure if that will affect things in unexpected ways. IMO this whole refactor is great but shouldn't be part of this PR, it can be another PR though.

File: dsp/modules/llama.py [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+    def __init__(
+        self,
+        model: str,
+        llama_model: Llama,
+        model_type: Literal["chat", "text"] = None,
+        **kwargs,
+    ):
+        super().__init__(model)
+
+        default_model_type = "chat" if "instruct" not in model else "text"
```
Comment (okhat) [inline dsp/modules/llama.py:25]: does this make sense here? this is for OpenAI models

File: dsp/modules/gpt3.py (hunk 4 above)
Comment (DanielUH2019) [reply dsp/modules/gpt3.py:160]: you are right, now I see that it is not equivalent, I will revert the refactor, and leave only the llama_cpp implementation.

File: dsp/modules/llama.py (hunk 5 above)
Comment (DanielUH2019) [reply dsp/modules/llama.py:25]: i will remove that to

File: dsp/modules/gpt3.py [hunk 6]
```diff
@@ -173,7 +173,7 @@ def __call__(
                     tokens, logprobs = tokens[:index], logprobs[:index]
 
                 avglog = sum(logprobs) / len(logprobs)
-                scored_completions.append((avglog, self._get_choice_text(c)))
+                scored_completions.append((avglog, self.get_choice_text(c)))
```
Comment (okhat) [inline dsp/modules/gpt3.py:176]: Hey I'd like to merge this PR, but I think this line is incorrect?

Comment (DanielUH2019) [reply dsp/modules/gpt3.py]: Yes, I didn't see that when reverting the refactoring; it's fixed now

File: dspy/__init__.py [hunk 7]
```diff
@@ -24,3 +24,5 @@
 
 Anyscale = dsp.Anyscale
 HFModel = dsp.HFModel
+
+LlamaCpp = dsp.LlamaCpp
```
Comment (okhat) [inline dspy/__init__.py:28]: This seems incorrect, since LlamaCpp is not imported by default?

File: dsp/modules/__init__.py [hunk 8]
```diff
@@ -8,4 +8,5 @@
 from .pyserini import *
 
 from .hf_client import HFClientTGI
-from .hf_client import Anyscale
\ No newline at end of file
+from .llama import LlamaCpp
```
Comment (okhat) [inline dsp/modules/__init__.py:11]: Hmm importing this by default means that people need a new dependency right? It should be optional

File: dsp/modules/__init__.py [hunk 9]
```diff
@@ -6,6 +6,7 @@
 from .cohere import *
 from .sbert import *
 from .pyserini import *
+from .llama import *
```
Comment (okhat) [inline dsp/modules/__init__.py:9]: Should we remove this line?

Comment (okhat) [reply dsp/modules/__init__.py:9]: Alternatively, we can move `from llama_cpp import Llama` into _inside_ `class LlamaCpp(LM)`.

What do you prefer?

Comment (DanielUH2019) [reply dsp/modules/__init__.py:9]: What would be the benefit of this? In the current implementation I think it follows the same structure as cohere

**Comment by arnavsinghvi11** [comment]
@DanielUH2019 just following up on old PRs :), Could you resolve relevant conflicts and lmk when ready to merge?

**Comment by codysnider** [comment]
@arnavsinghvi11 there's a new PR up as a continuation of this effort. it includes the snippet used to test against openhermes with llamacpp: https://github.com/stanfordnlp/dspy/pull/862

**Comment by arnavsinghvi11** [comment]
Closing this PR with #862 close to merging
